Not mandatory but may be useful :

Creating a txt file (or two) where one columns is all the contigs A and one column is all the contigs A

Name of the file : 'contigs_A_and_B.txt'

We just take the contigs in the 4th and 5th column of Ment_contigs_subgenomes without the NA.

We could create contigs_A.txt and contigs_B.txt and split them

Use the previous Jupyter Notebook to help

In [1]:
!pip install pandas

Defaulting to user installation because normal site-packages is not writeable


In [2]:
!pip install openpyxl

Defaulting to user installation because normal site-packages is not writeable


In [3]:
import pandas as pd
from openpyxl import Workbook

In [4]:
#Step 1: counting the number of assigned contigs

df=pd.read_excel('Ment_contigs_subgenomes.xlsx')

assigned_contigs=0
for index, row in df.iterrows():
    if row[3] == 'A' or row[3] == 'B': #The decision is the 4th column. The contig is assigned if it's A or B
        assigned_contigs+=1
        
print(assigned_contigs)
#Reminder we've got 556 contigs

192


We've got 192 contigs assigned.
220 contigs are in a block.
We've got a total of 556 contigs (source : PacBio)

33.81% of contigs are assigned.


In [5]:
#Step 2: counting the number of A and the number of B
#Verify if there are twice more A than B

df=pd.read_excel('Ment_contigs_subgenomes.xlsx')

contigs_A=0
contigs_B=0
for index, row in df.iterrows():
    if row[3] == 'A':
        contigs_A+=1
    elif row[3] == 'B':
        contigs_B+=1


print(contigs_A)
print(contigs_B)
contigs_A + contigs_B == assigned_contigs

143
49


True

We've got 143 contigs A and 49 contigs B. We've got more than 3x the numbers of contigs A than B.

It's not what we wanted because we wanted only 2x

In [7]:
#Step 3: we need to calculate the size of the contigs

#Step 3.1: get a dictionnary {contig:size}

file = open('contigs_length.txt', 'r')
lines = file.readlines()
file.close()

dic_contig_size={}
for i in range(0, len(lines), 2): #Only every two lines
    contig=lines[i].strip('>\n') #It supress the '>' and the '\n' of the line of contig
    size=lines[i + 1].strip()
    dic_contig_size[contig]=size
    
#print(dic_contig_size)

In [8]:
#Step 3.2: create the lists of contigs A and contigs B

contigs_list_A=[]
contigs_list_B=[]

df=pd.read_excel('Ment_contigs_subgenomes.xlsx')

for index, row in df.iterrows():
    #print(row[4])
    if row[3] == 'A' : #If the contig is an A
        contigs_list_A.append(row[4]) #This contig is added to the list of contig A
    elif row[3] == 'B' : #If the contig is a B
        contigs_list_B.append(row[5]) #This contig is added to the list of contig B
        
print(len(contigs_list_A)) #Excpected 143
print(len(contigs_list_B)) #Excpected 49
#print(contigs_list_A)

143
49


In [9]:
#Step 3.3 : calculate the size of the subgenomes A and B 
#by doing the sum of the size of all contigs A and the sum of the size of all contigs B

size_A=0
size_B=0
for c in dic_contig_size:
    if c in contigs_list_A:
        size_A+=int(dic_contig_size[c])
    elif c in contigs_list_B:
        size_B+=int(dic_contig_size[c])
        
print('size_A=',size_A)
print('size_B=',size_B)

size_A= 167341198
size_B= 55472704


In [10]:
size_A/size_B #Expected: arround 2

3.016640364241123

In [11]:
143/49

2.9183673469387754

In [12]:
#Supplemental task : create 3 files: contigs_A_ment.txt, contigs_B_ment.txt, contigs_A_and_B_ment.txt
df=pd.read_excel('Ment_contigs_subgenomes.xlsx')

#Creating file_A
file_A=open('contigs_A_ment.txt','w')
for index, row in df.iterrows():
    if row[3] == 'A' :
        file_A.write(row[4] + '\n')
file_A.close()
        
#Creating file_B
file_B=open('contigs_B_ment.txt','w')
for index, row in df.iterrows():
    if row[3] == 'B' :
        #print('On a un B')
        file_B.write(row[5] + '\n')
file_B.close()

In [ ]:
#paste the two files :
#do the following command lines in a terminal:

#paste contigs_A_ment.txt contigs_B_ment.txt > contigs_A_and_B_ment.txt
#echo -e "contigs_A\tcontigs_B" | cat - contigs_A_and_B_ment.txt > temp && mv temp contigs_A_and_B_ment.txt